network/src/noise/error.rs

```diff
@@ -0,0 +1,99 @@
+    MissingRemotePublicKey,
+
+    #[error("noise client: error building handshake init message: {0}")]
+    BuildClientHandshakeMessage(NoiseError),
+
+    #[error("noise client: error sending client handshake init message: {0}")]
+    ClientWrite(io::Error),
+
+    #[error(
+        "noise client: error reading server handshake response message, server \
+         probably rejected our handshake message: {0}"
+    )]
+    ClientRead(io::Error),
+
+    #[error("noise client: error finalizing secure connection: {0}")]
+    ClientFinalize(NoiseError),
+
+    #[error("noise server: error reading client handshake init message: {0}")]
+    ServerRead(io::Error),
+
+    #[error("noise server: client peer id is malformed somehow: {0}")]
+    InvalidClientPeerId(String),
+
+    #[error("noise server: detected self-dial: we're trying to connect to ourselves")]
+    SelfDial,
+
+    #[error("noise server: client {0}: client is expecting us to have the wrong public key: {1}")]
+    ClientExpectingWrongPubkey(ShortHexStr, String),
+
+    #[error("noise server: client {0}: error parsing handshake init message: {1}")]
+    ServerParseClient(ShortHexStr, NoiseError),
+
+    #[error("noise server: client {0}: lock is poisoned")]
+    LockPoisoned(ShortHexStr),
+
+    #[error(
+        "noise server: client {0}: known client connecting to us with \
+         unauthenticated public key: {1}"
+    )]
+    UnauthenticatedClientPubkey(ShortHexStr, String),
+
+    #[error("noise server: client {0}: client connecting with unauthenticated peer id: {1}")]
+    UnauthenticatedClient(ShortHexStr, PeerId),
+
+    #[error(
+        "noise server: client {0}: client's self-reported peer id and pubkey-derived peer \
+         id don't match: self-reported: {1}, derived: {2}"
+    )]
+    ClientPeerIdMismatch(ShortHexStr, PeerId, PeerId),
+
+    #[error("noise server: client {0}: handshake message is missing the anti-replay timestamp")]
+    MissingAntiReplayTimestamp(ShortHexStr),
+
+    #[error(
+        "noise server: client {0}: detected a replayed handshake message, we've \
+         seen this timestamp before: {1}"
+    )]
+    ServerReplayDetected(ShortHexStr, u64),
+
+    #[error("noise server: client {0}: error building handshake response message: {1}")]
+    BuildServerHandshakeMessage(ShortHexStr, NoiseError),
+
+    #[error("noise server: client {0}: error sending server handshake response message: {1}")]
+    ServerWrite(ShortHexStr, io::Error),
+
+    #[error("noise: {0}")]
+    Other(#[from] io::Error),
+}
+
+impl NoiseHandshakeError {
+    /// Errors that are either clear bugs or indicate some security issue. Should
+    /// immediately alert an engineer if we hit one of these errors.
+    // TODO(philiphayes): remove
+    #[allow(dead_code)]
+    pub fn should_security_log(&self) -> bool {
+        use NoiseHandshakeError::*;
+        match self {
+            MissingRemotePublicKey
+            | BuildClientHandshakeMessage(_)
+            | SelfDial
```

Comment:
why is this a security error?

Reply:
someone could have found a way to make you self dial, perhaps to DoS you